File: 0003_Longest_Substring_Without_Repeating_Characters/solution.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> char_map(128,-1);
        int left = 0;
        int max_len = 0;
        int n = s.length();

        for (int right = 0; right < n; right++) {
            char current_char = s[right];

            if (char_map[current_char] >= left) {
                left = char_map[current_char] + 1;
            }
            char_map[current_char] = right;

            int current_len = right - left + 1;
            if (current_len > max_len) {
                max_len = current_len;
            }
        }
        return max_len;
        
    }
};

File: 0003_Longest_Substring_Without_Repeating_Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_map = {}
        left = 0
        max_len = 0

        for right, char in enumerate(s):
            if char in char_map and char_map[char] >= left:
                left = char_map[char] + 1
                
            char_map[char] = right
            current_len = right - left + 1

            if current_len > max_len:
                max_len = current_len

        return max_len